config/tools/config_generator.cpp

```diff
@@ -0,0 +1,896 @@
+                      << ", does not exist. Attempting to copy it from its "
+                         "original location, "
+                      << temp_config_tools_dir.string() << " to "
+                      << temp_build_config_tools_dir.string() << std::endl;
+            copy_templates_to_build_dir();
+            // Try to use newly copied template files
+            std::string delimiter = "/";
+            std::string tmp_str = m_template_config_file;
+            size_t pos = 0;
+            std::string template_filename;
+            while((pos = tmp_str.find("/")) != std::string::npos) {
+                template_filename = tmp_str.substr(0, pos);
+                tmp_str.erase(0, pos + delimiter.length());
+            }
+            template_filename = tmp_str.substr(0, pos);
+            std::filesystem::path full_template_path_and_filename
+                = temp_build_config_tools_dir.append(template_filename);
+            if(std::filesystem::exists(full_template_path_and_filename)) {
+                m_template_config_file
+                    = full_template_path_and_filename.string();
+                template_file_is_valid = true;
+                std::cout << "Successfully copied " << template_filename
+                          << " from " << temp_config_tools_dir.string()
+                          << " to " << temp_build_config_tools_dir.string()
+                          << ". Using "
+                          << full_template_path_and_filename.string()
+                          << " as template file." << std::endl;
+            }
+        }
+    }
+
+    void config_generator::calculate_shard_coverage(const size_t num_shards,
+                                                    const size_t shard_size) {
+        std::vector<size_t> shard_index_sum_total(shard_size, 0);
+        shard_index_sum_total.reserve(shard_size);
+        // Setup shard metadata
+        for(size_t i = 0; i < num_shards; i++) {
+            size_t randNum = rand() % (shard_size - 0 + 1) + 0;
+            shard_info.push_back(ShardInfo());
+            shard_info.at(i).coverage = std::vector<size_t>(256, 0);
+            shard_info.at(i).coverage.at(randNum) = 1;
+            shard_info.at(i).still_expanding = true;
+            shard_info.at(i).allow_overlap = false;
+            shard_info.at(i).current_coverage_expansion_limits
+                = std::make_pair(randNum, randNum);
+            shard_info.at(i).numbers_covered = 1;
+            shard_info.at(i).shard_id = i;
+            auto overlap_percentage
+                = find_value<double>(tmpl_avg_shard_start_end_overlap_percent);
+            shard_info.at(i).overlap_percentage_allowed
+                = std::abs(calculate_normal_distribution_point(
+                    0,
+                    overlap_percentage.value()));
+            shard_index_sum_total.at(randNum)++;
+        }
+        bool still_expanding = true;
```

Comment:
There are a lot of places in this code where the variable type is unambiguous and clear based on the rhs of an assignment/initialization. Is there any reason to prefer spelling out the type over `auto`?

Reply:
Just for legibility. I usually try not to use auto since it also allows IDEs to more easily parse it and use autocompletion etc... 

Is there any advantage to using auto?

Reply:
Changed the bools and strings that could be to autos.

DONE

Reply:
> Is there any advantage to using auto?

It's fair to say this is largely a cosmetic stance. But, in my experience, fewer glyphs to parse = fewer places for mistakes to creep in (either in transcription or in comprehension). I also have spent a lot of time in much more heavy-powered type-inferring languages (e.g., Haskell), and the tendency to let the compiler do the book-keeping has definitely worked its way into my practice.